clippy_lints/src/needless_impls.rs

```diff
@@ -0,0 +1,109 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, get_parent_node, last_path_segment, ty::implements_trait};
+use rustc_errors::Applicability;
+use rustc_hir::{ExprKind, ImplItem, ImplItemKind, ItemKind, Node, UnOp};
+use rustc_hir_analysis::hir_ty_to_ty;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::EarlyBinder;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{sym, symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for manual implementations of `Clone` when `Copy` is already implemented.
+    ///
+    /// ### Why is this bad?
+    /// If both `Clone` and `Copy` are implemented, they must agree. This is done by dereferencing
+    /// `self` in `Clone`'s implementation. Anything else is incorrect.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// #[derive(Eq, PartialEq)]
+    /// struct A(u32);
+    ///
+    /// impl Clone for A {
+    ///     fn clone(&self) -> Self {
+    ///         Self(self.0)
+    ///     }
+    /// }
+    ///
+    /// impl Copy for A {}
+    /// ```
+    /// Use instead:
+    /// ```rust,ignore
+    /// #[derive(Eq, PartialEq)]
+    /// struct A(u32);
+    ///
+    /// impl Clone for A {
+    ///     fn clone(&self) -> Self {
+    ///         *self
+    ///     }
+    /// }
+    ///
+    /// impl Copy for A {}
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_CLONE_IMPL,
+    correctness,
+    "manual implementation of `Clone` on a `Copy` type"
+}
+declare_lint_pass!(NeedlessImpls => [NEEDLESS_CLONE_IMPL]);
+
+impl LateLintPass<'_> for NeedlessImpls {
+    #[expect(clippy::needless_return)]
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &ImplItem<'_>) {
+        let node = get_parent_node(cx.tcx, impl_item.hir_id());
+        let Some(Node::Item(item)) = node else {
+            return;
+        };
+        let ItemKind::Impl(imp) = item.kind else {
+            return;
+        };
+        let Some(trait_impl) = cx.tcx.impl_trait_ref(item.owner_id).map(EarlyBinder::skip_binder) else {
+            return;
+        };
```

Comment:
Wouldn't it be smarter, to use `check_item`, search for `ItemKind::Impl` and then walk downwards? Then you can do all of these checks only once. The lint would need to be emitted with [`span_lint_hir`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/diagnostics/fn.span_lint_hir.html#) to make sure that the lint level is tracked correctly. But the rest should work just fine :)

Reply:
I think this runs into the issue that it needs to check every item of `ItemKind::Impl`, rather than just one. I think that's how it was done but resulted in it being rather hard to read and a bit slow.

Reply:
Interesting, I would expect it to be faster, as we only check the elements inside impl items and avoid the lookup, but I might be mixing up projects rn xD. I believe we can keep it like this. I'll give you another review this week :)

Reply:
In truth, it's completely possible it is as I hadn't adequately benchmarked the old one. I'll reimplement the previous and see the proper difference

Reply:
I don't think that's necessary, but the results would be interesting :)

Reply:
I just benchmarked it, and it's ~2x slower than the current implementation, which is very surprising, I have no idea what's causing that.

I think I'll stick with the current implementation because of that

Reply:
SGTM, I'll give you a final review this week and most likely r+ it :)